students/john_navitsky/session06/mailroom.py

```diff
@@ -168,13 +182,52 @@ def print_thank_you(donor_id,hint="wonderful",dest=sys.stdout):
     print_lines(2,dest)
     print("-"*80,file=dest)
 
-    print(letter.format(hint,**donors[donor_id]),file=dest)
+    print(letter.format(hint,**current_donor),file=dest)
 
     print("-"*80,file=dest)
     print_lines(2,dest)
 
 
-def thank_you_entry():
+def match_donor(current_donor, donors):
+        """ See if record matches existing record, if so, return full record. """
+        for donor_id in donors:
+            existing_donor=donors[donor_id]
+            if existing_donor["full_name"].lower() == current_donor["full_name"].lower():
+                # if the owner exists, pull in the complete donor record
+                current_donor=existing_donor
+        return current_donor
+
+
+def update_donor(current_donor, donors, new_donation):
+        """ Update the donor database with new donation or donor record as needed. """
+
+        # match to existing record, if exists
+        current_donor = match_donor(current_donor, donors)
+
+        # datetime.datetime.strptime(<iso date>,'%Y-%m-%dZ')
+        today = datetime.datetime.utcnow().date().isoformat() + "Z"
+
+        # add donation to an existing donor
+        if "donor_id" in current_donor.keys():
```

Comment:
This is just `if donor_id in current_donor`

or handle this with:

```
try:
    donor_id = current_donor["donor_id"]
except KeyError
```

But why would you have a donor record without an id anyway? Better to always put in id in why you first create the donor record.

Reply:
This was a hold-over from when I had records from before the refactor so I was handling "updating" the contents.  At some point I decided migrating data wasn't a meaningful feature and I should have removed this.

That said, yes, I agree, I should make sure I create the id as soon as I know it isn't an existing record.